ext/MitsubaLoader/CMitsubaMaterialCompilerFrontend.cpp

```diff
@@ -0,0 +1,597 @@
+        params.image = std::move(_img);
+
+        return asset::ICPUImageView::create(std::move(params));
+    }
+    static core::smart_refctd_ptr<asset::ICPUImageView> createDerivMap(asset::ICPUImage* _heightMap, asset::ICPUSampler* _smplr)
+    {
+        const auto& sp = _smplr->getParams();
+        return createImageView(
+            createDerivMapFromHeightMap(
+                _heightMap, 
+                static_cast<asset::ICPUSampler::E_TEXTURE_CLAMP>(sp.TextureWrapU),
+                static_cast<asset::ICPUSampler::E_TEXTURE_CLAMP>(sp.TextureWrapV),
+                static_cast<asset::ICPUSampler::E_TEXTURE_BORDER_COLOR>(sp.BorderColor)
+            )
+        );
+    }
+
+    auto CMitsubaMaterialCompilerFrontend::getTexture(const CElementTexture* _element) const -> tex_ass_type
+    {
+        static_assert(std::is_same_v<tex_ass_type, SContext::tex_ass_type>, "These types must be same!");
+
+        auto found = m_loaderContext->textureCache.find(_element);
+        if (found == m_loaderContext->textureCache.end())
+            return tex_ass_type(nullptr, nullptr, 0.f);
+
+        return found->second;
+    }
+
+    asset::material_compiler::IR::INode* CMitsubaMaterialCompilerFrontend::compileToIRTree(asset::material_compiler::IR* ir, const CElementBSDF* _bsdf)
+{
+    using namespace asset;
+    using namespace material_compiler;
+
+    auto getFloatOrTexture = [this](const CElementTexture::FloatOrTexture& src, IR::INode::SParameter<float>& dst)
+    {
+        if (src.value.type == SPropertyElementData::INVALID)
+        {
+            dst.source = IR::INode::EPS_TEXTURE;
+            //make sure smart_refctd_ptr assignment ptr wont try to drop() -- .value is union
+            dst.value.constant = 0.f;
+            std::tie(dst.value.texture.image, dst.value.texture.sampler, dst.value.texture.scale) = getTexture(src.texture);
+            if (!dst.value.texture.image) {
+                assert(!dst.value.texture.sampler);
+                dst.source = IR::INode::EPS_CONSTANT;
+                dst.value.constant = 0.f;//0 in case when didnt find texture
+            }
+
+        }
+        else
+        {
+            dst.source = IR::INode::EPS_CONSTANT;
+            dst.value.constant = src.value.fvalue;
+        }
+    };
+    auto getSpectrumOrTexture = [this](const CElementTexture::SpectrumOrTexture& src, IR::INode::SParameter<IR::INode::color_t>& dst)
+    {
+        if (src.value.type == SPropertyElementData::INVALID)
+        {
+            dst.source = IR::INode::EPS_TEXTURE;
+            //make sure smart_refctd_ptr assignment ptr wont try to drop() -- .value is union
+            dst.value.constant = IR::INode::color_t(0.f);
+            std::tie(dst.value.texture.image, dst.value.texture.sampler, dst.value.texture.scale) = getTexture(src.texture);
+            if (!dst.value.texture.image) { //TODO should be using loader override here
+                assert(!dst.value.texture.sampler);
```

Comment:
?

Reply:
if image is null, it means it couldnt be loaded for some reason

Reply:
the place to use the loader override is not the compiler frontend but the Mitsuba XML loader/parser

Reply:
right, this comment is useless. It will wrok fine if user provides custom loader override. Resolved